File: openinstall-android/DeviceFingerprint.kt
package com.openinstall.tracking

import android.content.Context
import android.os.Build
import android.util.DisplayMetrics
import android.view.WindowManager
import org.json.JSONObject
import java.security.MessageDigest
import java.util.*

/**
 * 设备指纹收集工具类
 */
object DeviceFingerprint {
    
    /**
     * 收集设备指纹
     */
    fun collect(context: Context): JSONObject {
        val windowManager = context.getSystemService(Context.WINDOW_SERVICE) as WindowManager
        val displayMetrics = DisplayMetrics()
        windowManager.defaultDisplay.getMetrics(displayMetrics)
        
        return JSONObject().apply {
            put("userAgent", getUserAgent())
            put("platform", "Android")
            put("osVersion", Build.VERSION.RELEASE)
            put("deviceModel", Build.MODEL)
            put("deviceBrand", Build.BRAND)
            put("deviceManufacturer", Build.MANUFACTURER)
            put("screenWidth", displayMetrics.widthPixels)
            put("screenHeight", displayMetrics.heightPixels)
            put("screenDensity", displayMetrics.density)
            put("screenDensityDpi", displayMetrics.densityDpi)
            put("timezone", TimeZone.getDefault().id)
            put("timezoneOffset", TimeZone.getDefault().rawOffset / 1000 / 60) // 转换为分钟
            put("language", Locale.getDefault().language)
        }
    }
    
    /**
     * 获取 User Agent
     */
    private fun getUserAgent(): String {
        return "Android/${Build.VERSION.RELEASE} ${Build.MODEL}"
    }
    
    /**
     * 生成设备指纹ID
     */
    fun generateId(fingerprint: JSONObject): String {
        val jsonString = fingerprint.toString()
        val bytes = MessageDigest.getInstance("MD5").digest(jsonString.toByteArray())
        return bytes.joinToString("") { "%02x".format(it) }
    }
}

File: openinstall-flutter/lib/src/models/device_fingerprint.dart
/// 设备指纹信息
class DeviceFingerprint {
  final String? userAgent;
  final String? language;
  final String? platform;
  final int? screenWidth;
  final int? screenHeight;
  final int? screenColorDepth;
  final double? pixelRatio;
  final String? timezone;
  final int? timezoneOffset;
  final String? canvasFingerprint;
  final Map<String, dynamic>? webglFingerprint;
  final bool? cookieEnabled;
  final String? doNotTrack;
  
  // iOS/Android 特有字段
  final String? osVersion;
  final String? deviceModel;
  final String? deviceBrand;
  final String? deviceName;
  final double? screenScale;
  final double? screenDensity;

  DeviceFingerprint({
    this.userAgent,
    this.language,
    this.platform,
    this.screenWidth,
    this.screenHeight,
    this.screenColorDepth,
    this.pixelRatio,
    this.timezone,
    this.timezoneOffset,
    this.canvasFingerprint,
    this.webglFingerprint,
    this.cookieEnabled,
    this.doNotTrack,
    this.osVersion,
    this.deviceModel,
    this.deviceBrand,
    this.deviceName,
    this.screenScale,
    this.screenDensity,
  });

  Map<String, dynamic> toJson() {
    return {
      'userAgent': userAgent,
      'language': language,
      'platform': platform,
      'screenWidth': screenWidth,
      'screenHeight': screenHeight,
      'screenColorDepth': screenColorDepth,
      'pixelRatio': pixelRatio,
      'timezone': timezone,
      'timezoneOffset': timezoneOffset,
      'canvasFingerprint': canvasFingerprint,
      'webglFingerprint': webglFingerprint,
      'cookieEnabled': cookieEnabled,
      'doNotTrack': doNotTrack,
      'osVersion': osVersion,
      'deviceModel': deviceModel,
      'deviceBrand': deviceBrand,
      'deviceName': deviceName,
      'screenScale': screenScale,
      'screenDensity': screenDensity,
    };
  }

  factory DeviceFingerprint.fromJson(Map<String, dynamic> json) {
    return DeviceFingerprint(
      userAgent: json['userAgent'],
      language: json['language'],
      platform: json['platform'],
      screenWidth: json['screenWidth'],
      screenHeight: json['screenHeight'],
      screenColorDepth: json['screenColorDepth'],
      pixelRatio: json['pixelRatio']?.toDouble(),
      timezone: json['timezone'],
      timezoneOffset: json['timezoneOffset'],
      canvasFingerprint: json['canvasFingerprint'],
      webglFingerprint: json['webglFingerprint'] != null
          ? Map<String, dynamic>.from(json['webglFingerprint'])
          : null,
      cookieEnabled: json['cookieEnabled'],
      doNotTrack: json['doNotTrack'],
      osVersion: json['osVersion'],
      deviceModel: json['deviceModel'],
      deviceBrand: json['deviceBrand'],
      deviceName: json['deviceName'],
      screenScale: json['screenScale']?.toDouble(),
      screenDensity: json['screenDensity']?.toDouble(),
    );
  }
}


